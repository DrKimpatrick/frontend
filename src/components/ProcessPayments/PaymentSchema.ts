import * as Yup from 'yup';

const RequiredString = (msg: string) => Yup.string().trim().required(msg);

export const PaymentSchema = Yup.object().shape({
  billingAddress: RequiredString('Billing Address is required'),
  cardHolder: RequiredString('Cardholder is required')
});
