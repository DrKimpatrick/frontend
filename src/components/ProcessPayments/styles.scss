.stripe-input {
  border: 1px solid rgb(226, 232, 240);
  border-radius: 2px;
  padding: 12px;
  height: 47px;
  margin-top: 6px;
}

.process-payments-section {
  width: 33.3%;
}

@media (max-width: 786px) {
  .process-payments-section {
    width: 60%;

    .back-arrow {
      display: none;
    }

    .title {
      text-align: center;
      margin: auto;
    }
  }
}

@media (max-width: 500px) {
  .process-payments-section {
    width: 90%;
  }
}
