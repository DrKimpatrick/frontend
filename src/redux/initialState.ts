import users from 'redux/initialStates/users';

const initialState = {
  users
};

export default initialState;
