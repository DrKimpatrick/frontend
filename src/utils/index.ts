import jwt from 'jsonwebtoken';

type errorTypeRegister = {
    username: string;
    email: string;
    password: string;
}

type errorTypeLogin = {
    email: string;
    password: string;
}

type errorAddEmploy = {
    companyName : string;
    title: string;
    responsibilities: string;
    accomplishments: string;
    favoriteProject: string;
    supervisor: string;
    date: string;
    skillsUsed: string
}

export const validEmailRegex = RegExp(
    /^(([^<>()\[\]\.,;:\s@\"]+(\.[^<>()\[\]\.,;:\s@\"]+)*)|(\".+\"))@(([^<>()[\]\.,;:\s@\"]+\.)+[^<>()[\]\.,;:\s@\"]{2,})$/i
);

export const passwordRegex = RegExp(
    /^(?=.*[0-9])(?=.*[a-z])(?=.*[A-Z])(?=.*[@$!%*?&])([a-zA-Z0-9@$!%*?&]{8,})$/
)

export const validateForm = (errors:errorTypeRegister | errorTypeLogin | errorAddEmploy) => {
    let valid = true;
    Object.values(errors).forEach(val => val.length > 0 && (valid = false));
    return valid;
  };

export const decodeToken = (token: any) => {
    const decodedToken = jwt.decode(token)

    return decodedToken
}