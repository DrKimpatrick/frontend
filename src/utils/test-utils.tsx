import React, { FC, ReactElement } from 'react';
import { render } from '@testing-library/react';
import { SWRConfig } from 'swr';

const Wrapper: FC = ({ children }) => {
  return <SWRConfig value={{ dedupingInterval: 0 }}>{children}</SWRConfig>;
};

const customRender = (ui: ReactElement, options: any) =>
  render(ui, { wrapper: Wrapper, ...options });

// re-export everything
export * from '@testing-library/react';

// override render method
export { customRender as render };
