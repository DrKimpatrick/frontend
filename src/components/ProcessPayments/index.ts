import ProcessPayments from './ProcessPayments';

export { ProcessPayments };
