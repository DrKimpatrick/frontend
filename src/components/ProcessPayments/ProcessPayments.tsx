/* eslint-disable no-return-assign */
/* eslint-disable @typescript-eslint/camelcase */
import React, { useState } from 'react';
import useSWR from 'swr';
import { StripeCardElement, PaymentMethod } from '@stripe/stripe-js';
import { CardElement, useElements, useStripe } from '@stripe/react-stripe-js';
import ArrowBackTwoToneIcon from '@material-ui/icons/ArrowBackTwoTone';
import CircularProgress from '@material-ui/core/CircularProgress';
import { useLocation, useHistory } from 'react-router-dom';
import { Formik } from 'formik';

import axiosInstance from 'utils/axiosInstance';
import {
  Alert,
  AlertLevel,
  Field,
  MainBackground,
  NavBar,
  SplashScreen
} from 'components/Reusable';
import { ErrorMessageDialog } from 'components/Reusable/ErrorMessageDialog';

import './styles.scss';
import { PaymentSchema } from './PaymentSchema';
import { AdminsProcess, TalentProcess } from '../../redux/action-types/user';
import { Routes } from '../../utils/routes';

const ProcessPayments = () => {
  const elements = useElements();
  const stripe = useStripe();
  const [paymentLoading, setPaymentLoading] = useState(false);
  const [paymentFailed, setPaymentFailed] = useState(false);
  const [cardError, setCardError] = useState<string | undefined>('');
  const [pageAlert, setPageAlert] = useState<{
    message: string | undefined;
    show: boolean;
    level?: AlertLevel;
  }>({ message: '', show: false, level: AlertLevel.ERROR });
  const { data: userData } = useSWR('/users/me');
  const {
    state: { plan: productPlan, featureChoice, coupon, subsidy, successPage }
  } = useLocation<any>();
  const history = useHistory();

  const handleAlertClose = (event?: React.SyntheticEvent, reason?: string) => {
    if (reason === 'clickaway') {
      return;
    }

    setPageAlert(prev => ({ ...prev, show: false, level: AlertLevel.ERROR }));
  };

  const createSubscription = async (
    paymentMethod: PaymentMethod,
    customerName: string
  ) => {
    const subscriptionPayload = {
      customerInfo: {
        name: customerName,
        email: userData.profile.email,
        planId: productPlan.id
      },
      paymentMethodId: paymentMethod.id,
      featureChoice,
      profileProcess: subsidy
        ? AdminsProcess.Completed
        : TalentProcess.Completed,
      coupon,
      subsidy
    };

    const { data } = await axiosInstance.post(
      '/stripe/subscription',
      subscriptionPayload
    );

    manageSubscriptionStatus(data.subscription);
  };

  const manageSubscriptionStatus = async (subscription: any) => {
    const { latest_invoice } = subscription;
    const { payment_intent } = latest_invoice;

    if (stripe && payment_intent) {
      const { client_secret, status } = payment_intent;
      if (
        status === 'requires_action' ||
        status === 'requires_payment_method'
      ) {
        stripe
          .confirmCardPayment(client_secret)
          .then(result => {
            if (result.error) {
              setPageAlert({
                message: result.error.message,
                show: true,
                level: AlertLevel.ERROR
              });
            } else {
              setPageAlert({
                message: 'Your payment was succesful',
                show: true,
                level: AlertLevel.SUCCESS
              });
              if (successPage) {
                return (window.location.href = successPage);
                // return history.replace(successPage);
              }
              window.location.href = Routes.UserDashboard;
              // history.replace(Routes.UserDashboard);
            }
          })
          .catch((confirmationError: any) => {
            setPaymentFailed(true);
            setPageAlert({
              message:
                confirmationError.message ?? 'Payment confirmation failed',
              show: true,
              level: AlertLevel.ERROR
            });
          })
          .finally(() => {
            setPaymentLoading(false);
          });
      } else {
        setPageAlert({
          message: 'Your payment was succesful',
          show: true,
          level: AlertLevel.SUCCESS
        });
        if (successPage) {
          return (window.location.href = successPage);
          // return history.replace(successPage);
        }
        window.location.href = Routes.UserDashboard;
        // history.replace(Routes.UserDashboard);
      }
    } else {
      setPageAlert({
        message: 'Your payment was succesful',
        show: true,
        level: AlertLevel.SUCCESS
      });
      if (successPage) {
        return (window.location.href = successPage);
        // return history.replace(successPage);
      }
      window.location.href = Routes.UserDashboard;
      // history.replace(Routes.UserDashboard);
    }
  };

  const handlePayment = async (values: {
    billingAddress: string;
    cardHolder: string;
  }) => {
    setPaymentLoading(true);

    if (!stripe || !elements) {
      // eslint-disable-next-line no-useless-return
      return;
    }

    const cardElement = elements.getElement(CardElement) as StripeCardElement;
    const {
      error: paymentError,
      paymentMethod
    } = await stripe.createPaymentMethod({
      type: 'card',
      card: cardElement,
      billing_details: {
        email: userData.profile.email,
        address: {
          line1: values.billingAddress
        }
      }
    });

    if (paymentError || !paymentMethod) {
      setPaymentLoading(false);
    } else {
      createSubscription(paymentMethod, values.cardHolder);
    }
  };

  if (!stripe || !userData) {
    return <SplashScreen />;
  }

  return (
    <>
      <NavBar />
      <section className="process-payments-section mx-auto text-gray-texts">
        <Alert
          message={pageAlert.message}
          show={pageAlert.show}
          severity={pageAlert.level}
          onClose={handleAlertClose}
        />
        <div className="flex flex-col">
          <div className="text-2xl flex items-center">
            <ArrowBackTwoToneIcon
              className="-ml-12 mr-6 cursor-pointer back-arrow"
              onClick={() => history.goBack()}
            />
            <h1 className="font-bold title">
              Please provide your payment details
            </h1>
          </div>
          {paymentFailed ? (
            <ErrorMessageDialog
              title="Sorry, it looks like we are unable to process your payment"
              message="Please contact us or try again"
            />
          ) : (
            <Formik
              initialValues={{ billingAddress: '', cardHolder: '' }}
              validationSchema={PaymentSchema}
              onSubmit={handlePayment}
            >
              {({
                values,
                errors,
                touched,
                handleChange,
                handleBlur,
                handleSubmit
              }) => (
                <form
                  className="mt-4 w-full flex flex-col"
                  onSubmit={handleSubmit}
                >
                  <div className="mb-4">
                    <label className="mb-2 text-gray-600 text-sm">
                      Cardholder Name
                    </label>
                    <Field
                      type="text"
                      classes="mt-2"
                      value={values.cardHolder}
                      onChange={handleChange('cardHolder')}
                      onBlur={handleBlur('cardHolder')}
                      touched={touched.cardHolder}
                      error={errors.cardHolder}
                    />
                  </div>

                  <div className="mb-4">
                    <label className="mb-2 text-gray-600 text-sm">
                      Billing Address
                    </label>
                    <Field
                      type="text"
                      classes="mt-2"
                      value={values.billingAddress}
                      onChange={handleChange('billingAddress')}
                      onBlur={handleBlur('billingAddress')}
                      touched={touched.billingAddress}
                      error={errors.billingAddress}
                    />
                  </div>

                  <div>
                    <label className="mb-2 text-gray-600 text-sm">
                      Billing Details
                    </label>
                    <CardElement
                      className="stripe-input"
                      onChange={e => setCardError(e.error?.message)}
                    />
                    <span>{cardError}</span>
                  </div>

                  <button
                    type="submit"
                    className="mt-4 px-4 py-2 bg-green-700 text-white text-center self-center rounded-sm"
                    disabled={paymentLoading}
                  >
                    Process Payment{' '}
                    {paymentLoading && (
                      <span className="ml-2">
                        <CircularProgress color="inherit" size={20} />
                      </span>
                    )}
                  </button>
                </form>
              )}
            </Formik>
          )}
        </div>
      </section>
      <MainBackground />
    </>
  );
};

export default ProcessPayments;
