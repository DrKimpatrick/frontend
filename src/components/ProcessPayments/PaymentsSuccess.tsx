import React from 'react';
import ArrowBackTwoToneIcon from '@material-ui/icons/ArrowBackTwoTone';
import { useHistory } from 'react-router-dom';
import { MainBackground, NavBar } from 'components/Reusable';
import successCheck from 'assets/images/2-check.png';
import { Image, TrendingFlat } from '@material-ui/icons';

import './styles.scss';
import {Routes} from "../../utils/routes";

const PaymentsSuccess = () => {
  const history = useHistory();

  const saveAndReviewProfile = () => {
    // TODO: not sure what is supposed to be saved(the mockups have different ideas)
    history.replace(Routes.UserDashboard);
  };

  const reviewAndPublishProfile = () => {
    // TODO: not sure what is supposed to be saved(the mockups have different ideas)
    history.replace(Routes.UserDashboard);
  };

  return (
    <>
      <NavBar />
      <section className="w-1/3 mx-auto text-gray-texts">
        <div className="flex flex-col">
          <div className="text-2xl flex items-center">
            <ArrowBackTwoToneIcon
              className="-ml-12 mr-6 cursor-pointer"
              onClick={() => history.goBack()}
            />
            <h1 className="font-bold">Payment Successful</h1>
          </div>

          <div className="flex items-center justify-center flex-col mt-4 mb-4">
            <img src={successCheck} alt="Success Check" />
            <p className="text-center mt-2">
              <strong className="text-green-500">Congratulations</strong> you
              are a <b>premium</b> <br /> Tech Talent Member
            </p>
          </div>

          <div className="flex items-center justify-center flex-col mt-4 mb-4">
            <button
              disabled
              type="button"
              className="bg-gray-200 py-3 w-2/3 flex flex-row rounded justify-center"
            >
              <Image /> Upload Photo
            </button>
          </div>

          <div className="flex items-center justify-center flex-col mt-4 mb-4">
            <h4 className="font-bold mb-2">Profile Tagline</h4>
            <span>
              <strong>
                I engineer efficient solutions to complicated problems
              </strong>
            </span>
          </div>

          <div className="flex items-center justify-center flex-col mt-4 mb-4 space-y-5">
            <button
              type="button"
              className="bg-green-600 text-white py-3 w-2/3 flex flex-row rounded justify-around py-3"
              onClick={reviewAndPublishProfile}
            >
              Review & Publish Profile <TrendingFlat />
            </button>

            <button
              type="button"
              className="bg-green-600 text-white py-3 w-2/3 flex flex-row rounded justify-around py-3"
              onClick={saveAndReviewProfile}
            >
              Save & Review later <TrendingFlat />
            </button>
          </div>
        </div>
      </section>
      <MainBackground />
    </>
  );
};

export default PaymentsSuccess;
