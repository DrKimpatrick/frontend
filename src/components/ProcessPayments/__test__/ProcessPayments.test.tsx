import React from 'react';
import { MemoryRouter as Router } from 'react-router-dom';
import { render, cleanup } from 'utils/test-utils';
import { cache } from 'swr';
import { Provider } from 'react-redux';
import thunk from 'redux-thunk';
import configureMockStore from 'redux-mock-store';
import { Elements } from '@stripe/react-stripe-js';

import ProcessPayments from '../ProcessPayments';
import initialState from 'redux/initialState';
import { mockElement, mockElements, mockStripe } from '../__mocks__';

jest.mock('react-router-dom', () => ({
  ...jest.requireActual('react-router-dom'),
  useLocation: () => ({
    state: {
      plan: {
        id: 'some-unique-id'
      }
    }
  })
}));

jest.mock('@stripe/react-stripe-js', () => {
  const stripe = jest.requireActual('@stripe/react-stripe-js');

  return {
    ...stripe,
    Element: () => {
      return mockElement;
    },
    useStripe: () => {
      return mockStripe;
    },
    useElements: () => {
      return mockElements;
    }
  };
});

describe(ProcessPayments, () => {
  let store: any;
  let stripe: any;
  let stripePromise: any;
  let mockElements: any;
  const middleware = [thunk];
  const mockStore = configureMockStore(middleware);

  beforeEach(() => {
    store = mockStore(initialState);
    stripe = mockStripe();
    stripePromise = Promise.resolve(stripe);
    mockElements = Symbol('MockElements');
    stripe.elements.mockReturnValue(mockElements);
  });

  afterEach(() => {
    cache.clear();
    cleanup();
  });

  it('matches the snapshot', () => {
    const tree = render(
      <Provider store={store}>
        <Elements stripe={stripe}>
          <Router>
            <ProcessPayments />
          </Router>
        </Elements>
      </Provider>,
      {}
    );
    expect(tree).toMatchSnapshot();
  });
});
