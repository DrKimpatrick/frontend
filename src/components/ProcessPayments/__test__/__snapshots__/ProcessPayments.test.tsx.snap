// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProcessPayments matches the snapshot 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="splashScreen"
      >
        <div
          class="logo"
        >
          <img
            alt="logo"
            src="logo-image.png"
          />
        </div>
        <div
          class="loading"
        >
          <h5>
            Loading
          </h5>
          <div
            class="css-10q6n6d"
          >
            <div
              class="css-14wk343"
            />
            <div
              class="css-13zqm2w"
            />
            <div
              class="css-14wk343"
            />
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="splashScreen"
    >
      <div
        class="logo"
      >
        <img
          alt="logo"
          src="logo-image.png"
        />
      </div>
      <div
        class="loading"
      >
        <h5>
          Loading
        </h5>
        <div
          class="css-10q6n6d"
        >
          <div
            class="css-14wk343"
          />
          <div
            class="css-13zqm2w"
          />
          <div
            class="css-14wk343"
          />
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
